File: frontend/lib/models/course.dart
class Course {
  final String? id;
  final String courseCode;
  final String courseName;
  final int credits;
  final List<String>? branches;
  final String? primaryRoom;

  Course({
    this.id,
    required this.courseName,
    required this.courseCode,
    required this.credits,
    this.branches,
    this.primaryRoom,
  });

  factory Course.fromJson(Map<String, dynamic> json) {
    return Course(
      id: json['_id'],
      courseCode: json['courseCode'],
      courseName: json['name'],
      credits: json['credits'],
      branches:
          (json['branches'] as List).map((item) => item as String).toList(),
      primaryRoom: json['primaryRoom'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'courseCode': courseCode,
      'name': courseName,
      'credits': credits,
      'branches': branches,
      'primaryRoom': primaryRoom,
    };
  }
}

File: backend/models/course.js
import mongoose from "mongoose";

const courseSchema = new mongoose.Schema(
  {
    name: { type: String, required: true, trim: true },
    courseCode: {
      type: String,
      required: true,
      unique: true,
      trim: true,
      uppercase: true, // Ensures course codes like "CS101" remain standardized
      index: true, // Speeds up course lookups
    },
    primaryRoom: { type: String, trim: true },
    credits: { type: Number, required: true, min: 1 }, // Ensures credits are positive
    professorId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Faculty",
      required: true,
    },
    branches: [{ type: String, required: true, trim: true }], // Prevents empty values in array
  },
  { timestamps: true } // Automatically adds createdAt & updatedAt
);

const Course = mongoose.model("Course", courseSchema);
export default Course;
